### ISSUE TITLE
How to read big (20GB) compressed json arrays?

### ISSUE THREAD

**Issue by tadas-subonis** [dask/dask#2423]
Hi,

I am trying to read a bunch of JSON objects (and filter them out into another file) that are compressed in one 20GB gzip compressed file that are delimited by newline characters.

I've tried:
```
    filename = 'feed_entries.jsona.gz'
    with gzip.open(filename) as lines:
        feed_entries = (
            db.from_sequence(lines)
```
but this "from_sequence" wants to convert entire sequence to a list (which exhausts all of my available memory)

The read file approach doesn't work either:
```
db.read_text(filename)
                .map(str.strip)
                .map(loads)
                .filter(lambda fe: something)
                .filter(lambda fe:something)
                .to_textfiles('out\\*.jsona.gz')
```

as it first tries to read entire dataset into memory.

Any advice?

**Comment by mrocklin** [comment]
I would expect the latter option to work.  Note that it wouldn't parallelize significantly because gzip doesn't offer random-access reading (you have to stream through the whole thing linearly), but it should handle everything in a streaming/low-memory fashion.

It looks like graph optimization isn't kicking in.  I have a reproducible test case on my end. I'll take a look.

**Comment by mrocklin** [comment]
Hrm, actually it looks like this is unlikely to work.  Ever since we moved to supporting more bytes backends we had to remove streaming writes.  I suspect that if you were to downgrade dask significantly that things would work for you.  

cc @martindurant any thoughts on how to write out iterators with dask.bag using the new write_bytes machinery?  Do we pass in file objects?

**Comment by martindurant** [comment]
[This code branch](https://github.com/dask/dask/blob/master/dask/bytes/core.py#L44) is specifically for iterable things - but the whole sequence to be written is already in memory at this point. bag.to_textfiles does use bytes.write_bytes, and there are tests for this.

(It may be true that using straight gz tools to first split the file into several un-compressed filed is not a bad idea)

**Comment by mrocklin** [comment]
> (It may be true that using straight gz tools to first split the file into several un-compressed filed is not a bad idea)

Good point.  @tadas-subonis your life would be much better if you would decompress first.  This would also allow dask to parallelize your computation.

**Comment by mrocklin** [comment]
> This code branch is specifically for iterable things - but the whole sequence to be written is already in memory at this point. bag.to_textfiles does use bytes.write_bytes, and there are tests for this.

OK, so I can send a giant iterator to `write_block` and things will work smoothly?

**Comment by martindurant** [comment]
>  I can send a giant iterator

That's the idea, although the tests are not giant.

**Comment by mrocklin** [comment]
OK, then I've resolved this issue in #2425

**Comment by jcrist** [comment]
This was fixed by #2425. Closing.
